`generate` and machine allocation bug fixes

FrederoxDev opened the pull request:
Fixes:
- fixed buggy implementation of returning power to generators
- reduced repeated code by introducing a helper func
- changed how generate works minorly under the hood, no breaking changes
- fixed race condition that lead to infinite power when two storages were on the same network
- makes generate actually work as described in the issue below.

Related Issue: https://github.com/Fluffyalien1422/bedrock-energistics-core/issues/68 (already been closed early)

FrederoxDev commented:
Just added an optional param to generate `useReserveStorage` that is true by default (same behaviour as before), when set to false it will disable using reserve storage, allowing the machine to control exactly what is outputted.

FrederoxDev commented:
Any update? I'm hoping to get this merged because it has a fix that allows my pack to work && id like to start implementing conditional sides and don't want to be waiting on PRs

Fluffyalien1422 commented:
> Any update? I'm hoping to get this merged because it has a fix that allows my pack to work && id like to start implementing conditional sides and don't want to be waiting on PRs

I was focusing on releasing v0.5.0, I'll review this soon.

Fluffyalien1422 commented on packs/BP/scripts/network_ipc.ts:
```diff
@@ -106,19 +106,24 @@ export function generateListener(payload: ipc.SerializableValue): null {
 
-  const storageType = InternalRegisteredStorageType.forceGetInternal(type);
+  // Set the new amount, as *send will take it directly out of storage
+  setMachineStorage(block, type, newAmount);
```
What if `newAmount` is greater than the machine's max storage?

Fluffyalien1422 replied:
nothing is added if `consumeExisting` is `true`, so we only need to call `setMachineStorage` if `consumeExisting` is `false`
```suggestion
  if (!consumeExisting) {
    setMachineStorage(block, type, newAmount);
  }
```

Fluffyalien1422 commented on packs/BP/scripts/network.ts:
```diff
@@ -102,19 +102,36 @@ export class MachineNetwork extends DestroyableObject {
+      // default behaviour of generate is to send the amount to generate + the current in the machine
+      // Instead do it by send.amount allowing for machines to use an alternative function to only output
+      // the amount generated, and not stored inside the machine always.
+      setMachineStorage(
```
This seems like it could cause some weird issues. We don't know how long it's going to take until the last `sendGroupAllocation` is called. With the old system, the generators storage was set once rather than subtracting then waiting then adding again.

Fluffyalien1422 commented:
Okay, I'm still confused as to the purpose of this PR.
> fixed buggy implementation of returning power to generators
I agree that the previous implementation was bad, but this implementation is pretty much the same with extra steps.
> fixed race condition that lead to infinite power when two storages were on the same network
This was never a problem. Your original issue (#68) was caused by an incorrect usage of `generate`, as I explained in [my response](https://github.com/Fluffyalien1422/bedrock-energistics-core/issues/68#issuecomment-2486883348).
> makes generate actually work as described in the issue below.
`generate` did work as described. Can you explain what exactly you changed? I see you added two new arguments, and other than that did not change how `generate` works by default.

Fluffyalien1422 commented:
> Any update? I'm hoping to get this merged because it has a fix that allows my pack to work

What exactly did this fix in your pack?

FrederoxDev commented:
The original fix led to low priority storages instantly completely dropping all stored storage and broke power returning to them. Pretty critical because it completely breaks my pack. 

Also this fix, fixed the fact that power was never getting consumed from machines which was the behaviour you described in that issue

I'm quite confused as to the level of pushback I'm getting from this PR? I thought this would be a simple merge and problem solved xD

Fluffyalien1422 commented:
> Also this fix, fixed the fact that power was never getting consumed from machines which was the behaviour you described in that issue

Okay, I just figured out what you meant. When a machine already has energy and multiple generators on the network, it won't remove the energy.

Fluffyalien1422 commented on packs/BP/scripts/network_ipc.ts:
```diff
@@ -106,19 +106,24 @@ export function generateListener(payload: ipc.SerializableValue): null {
 
-  const fullAmount = amount + getMachineStorage(location, type);
-  if (!fullAmount) return null;
+  // If consume existing is true, do not add new power, instead consume the existing storage
```
```suggestion
  // If consume existing is true, do not increase the storage, instead consume the existing storage
```

Fluffyalien1422 commented on packs/BP/scripts/network_ipc.ts:
```diff
@@ -106,19 +106,24 @@ export function generateListener(payload: ipc.SerializableValue): null {
   const type = data.b;
   const amount = data.c;
+  const useReserveStorage = data.d;
+  const consumeExisting = data.e;
```
`consumeExisting` should be removed or made into a different function, as it doesn't make sense here. If `useReserveStorage` and `consumeExisting` are both `true`, then `amount` (a required argument) is never used.

FrederoxDev replied:
issue resolved by using an enum called mode instead of indivudal parameters to control each thing

FrederoxDev commented:
I believe I have now resolved this issue, I was able to replicate it with a large number of machines in my own pack && i have been able to test that it is fixed in your repro pack. I also changed my approach with the changes to generate, instead I changed it to be an enum which chooses what was before those individual parameters:
```
export declare enum GenerateMode {
    /**
     * - Sends existing reserve storage
     * - Adds the amount generated ontop of existing reserve
     */
    Default = 0,
    /**
     * - Only sends the amount passed in generate, does not include reserve
     * - Takes the amount away from existing reserve
     */
    TakeFromReserve = 1
}
```

Fluffyalien1422 commented:
There is still an issue where the storage values flash (change from 0 to another value constantly) because `generate` sets the machine's storage to 0, then some time passes before `send` sets the storage to the correct value. However, I was able to fix the original issue (#70) in commit b710948079c3cb092eaa1fe4fab06323f6e9da5a

Fluffyalien1422 commented:
Since the original issue has been resolved, is this PR now just adding new features to `generate` and cleaning up the code in `send`? Please update the title and description to reflect the current state of this PR. I'm going to set this as a draft for now.

FrederoxDev commented:
If you've fixed the implementation, ill probably end up closing this PR and adding in my other generate functionality later on, currently going on a UI stint so I won't be working on this side of my pack until the weekend probably.
